### DISCUSSION TITLE
Mirror Max Width values for Min Width

### DISCUSSION THREAD

**Discussion by phacks** [tailwindlabs/tailwindcss#1917] [Ideas]
I found myself needing to set a minimum width for a component, and mechanically wrote `min-w-48` as I thought that any valid Tailwind max width value would have its counterpart in the min width classes.

Turns out that min-width only supports two values: `min-w-0` and `min-w-full` whereas max-width has 16.

I know the Tailwind config can be extended for my use case, but I’d like to suggest the idea of adding a few more min-width classes, to approach parity between both:

Class | Properties
-- | --
.min-w-xs | min-width: 20rem;
.min-w-sm | min-width: 24rem;
.min-w-md | min-width: 28rem;
.min-w-lg | min-width: 32rem;
.min-w-xl | min-width: 36rem;
.min-w-2xl | min-width: 42rem;
.min-w-3xl | min-width: 48rem;
.min-w-4xl | min-width: 56rem;
.min-w-5xl | min-width: 64rem;
.min-w-6xl | min-width: 72rem;
.min-w-full | min-width: 100%;
.min-w-screen-sm | min-width: 640px;
.min-w-screen-md | min-width: 768px;
.min-w-screen-lg | min-width: 1024px;
.min-w-screen-xl | min-width: 1280px;
.min-w-0 | min-width: none;

(this issue can apply to `min-height` too).

Happy to work on a PR if y’all think this is worth it!

**Comment by adamwathan** [comment]
Hey! Can you explain the situation you ran into where you needed this? I’ve not run into one yet myself 🤔

**Comment by phacks** [comment]
Sure! I have a custom select box with options inside of varying length, eg. `cats`, `dogs`, `hummingbird`, `some very long animal name`.

I want to set a min-width on the selectbox so that:

1. If the selected option is very long (like the last one), the width adapts to it;
2. If the selected option is very short (like the first two ones), the width doesn’t shrink all the way down but stays at a minimum value (which would fit “hummingbird”)

I’m trying to achieve this to avoid moving things around the screen too much when the user changes from a long option to a short one.

This is admittedly a niche problem 😅

I’m using Tailwind UI’s free preview select box by the way (I love it!)—here’s what I’m trying to mitigate:

![CleanShot 2020-06-12 at 11 43 00](https://user-images.githubusercontent.com/2587348/84489450-0a8e5180-aca2-11ea-872d-b2a6e44d8e90.gif)

**Comment by estevanmaito** [comment]
I thought about it yesterday, but realised that my case could be too specific, but here it is another example where this could be helpful.

Creating notification badges, the only way to achieve a perfect rounded badge (first highlighted code) is to apply fixed dimensions (`min-width` would be ideal). But then, if we get more than one character inside the badge, it won't grow, so the solution is to use padding utilities, that doesn't look good with one digit, but grows ok (second highlight).

![badges](https://user-images.githubusercontent.com/8527573/84510700-d9128780-ac9b-11ea-93be-66247d00de3f.png)

You can see it live (and the code for it) here: https://tailwind-starter-kit.now.sh/docs/badges

**Comment by msheakoski** [comment]
I was looking for this today for a sidebar. I wanted to make it a minimum width but expand as needed depending on the content.

Similarly, for forms that have labels on the left, it is nice to keep a minimum width to visually separate them from the fields, with room to grow instead of wrapping the label text.

Also, some modals / dialogs might look funny if they are too skinny, so setting a min/max width range would be nice.

**Comment by estevanmaito** [comment]
Another case: height animation.

```html
<ul
  x-transition:enter="transition-all ease-in-out duration-300"
  x-transition:enter-start="opacity-0 max-h-0"
  x-transition:enter-end="opacity-100 max-h-xl"
  x-transition:leave="transition-all ease-in-out duration-300"
  x-transition:leave-start="opacity-100 max-h-xl"
  x-transition:leave-end="opacity-0 max-h-0"
>
```

![translate-height](https://user-images.githubusercontent.com/8527573/84578230-d433fd80-ad99-11ea-8400-2533d1a395d1.gif)

Also, note that the last line in OP's propose says `.min-w-0 | min-width: none;`, but `none` is only applied to `max-height/width` properties, and the current behavior of the existing `min-w-0` is right.

Implementing `max-h-0` as my example above, it would be needed to create the same class for `max-w` utilities, as there is no 0 there.

**Comment by estevanmaito** [comment]
Another case: missing `min-w-max-content` which is implemented in Tailwind UI

`w-56`
![Screenshot_2020-06-21 React App](https://user-images.githubusercontent.com/8527573/85230028-c139b280-b3c3-11ea-8ad2-3aac1e458e39.png)

`w-56` with more text
![Screenshot_2020-06-21 React App(2)](https://user-images.githubusercontent.com/8527573/85230057-ec240680-b3c3-11ea-8f0e-f2a8c5122a00.png)

`w-56` with more text and `min-w-max-content`
![Screenshot_2020-06-21 React App(1)](https://user-images.githubusercontent.com/8527573/85230029-c1d24900-b3c3-11ea-8ade-2ecb5cc753fc.png)

Obviously the other variants should also be added: `min-w-max-content`, `max-w-max-content`, ...

**Comment by phacks** [comment]
@adamwathan Happy to work on a Pull Request for this if you think it has value for Tailwind!

**Comment by adamwathan** [comment]
Can't decide if we should add this or not by default. Will keep this open for now, thanks for the suggestion 👍

**Comment by robertdrakedennis** [comment]
regardless, this would be an excellent external plugin and not everything needs to be added by default

**Comment by hrafnkellpalsson** [comment]
I was making a button that either displays a text ("Place order") or a spinner and the button's width would change depending on which one it was displaying (since I was just adding padding around the children content of the button).

My first thought was that I needed to set a min width and was surprised I didn't find it in the docs. Honestly am not good enough at CSS to know if this belongs in the framework or not but maybe the docs should explain why it's not there and alternative strategies?

**Reply by estevanmaito** [reply]
I think in your case, for a button, the best option and most consistent, in my opinion, is to use padding and not margin.

**Reply by hrafnkellpalsson** [reply]
Oh sorry maybe I was being clear enough but I was using padding on the button, no margin's involved.

**Comment by jameswoodley** [comment]
If anybody wants the variant styles to save a typing out:

```
@layer utilities {
        @variants responsive {
            .max-w-4xs          { max-width: 8rem; }
            .max-w-3xs          { max-width: 12rem; }
            .max-w-2xs          { max-width: 16rem; }

            .min-w-0             { min-width: 0rem; }
            .min-w-none       { min-width: none; }
            .min-w-4xs          { min-width: 8rem; }
            .min-w-3xs          { min-width: 12rem; }
            .min-w-2xs          { min-width: 16rem; }
            .min-w-xs	        { min-width: 20rem; }
            .min-w-sm	        { min-width: 24rem; }
            .min-w-md	        { min-width: 28rem; }
            .min-w-lg	        { min-width: 32rem; }
            .min-w-xl	        { min-width: 36rem; }
            .min-w-2xl	        { min-width: 42rem; }
            .min-w-3xl	        { min-width: 48rem; }
            .min-w-4xl	        { min-width: 56rem; }
            .min-w-5xl	        { min-width: 64rem; }
            .min-w-6xl	        { min-width: 72rem; }
            .min-w-7xl	        { min-width: 80rem; }
            .min-w-full	        { min-width: 100%; }
            .min-w-min	        { min-width: min-content; }
            .min-w-min	        { min-width: min-content; }
            .min-w-prose        { min-width: 65ch; }
            .min-w-screen-sm    { min-width: 640px; }
            .min-w-screen-md	{ min-width: 768px; }
            .min-w-screen-lg	{ min-width: 1024px; }
            .min-w-screen-xl	{ min-width: 1280px; }
            .min-w-screen-2xl	{ min-width: 1536px; }
        }
    }
```

I added some lower ones to not stop at 20rem

**Comment by edman** [comment]
Can someone explain why min-w-X is not as useful as the max-w-X counterpart?

I expected them to have the same (analogous) use: something can have a variable width, max-w limits the upper range, min-w limits the lower range.

I'm a CSS noob, apologies if this is obvious to others.

**Comment by binaryartifex** [comment]
yeah just wanted to pipe in and stoke this fire as well. im using this all over the place in both work and professional projects, im forced to do things like `min-w-[320px]` for something like a modal, the select drop down has already been mentioned, etc. I think its definitely a greater value add to include these utilities in the min-width utility (and a little more consistent imo) as opposed to essentially getting the reference to the width i need just to hard code it each time

**Comment by mikuhl-dev** [comment]
I am confused why this is not included?

**Comment by r-thomson** [comment]
Is there a good reason that `min`/`max-w`/`h` shouldn't support same set of units as `w-` and `h-`? Perhaps there was something to be said about file size in the past, but now that Tailwind is JIT by default that's not a concern.

**Reply by adamwathan** [reply]
Yeah was file-size motivated in the past and just haven't made it a priority to revisit this since JIT. On my list of things to do for v4 though 👍

**Reply by mrleblanc101** [reply]
So if i have `w-9` which I want to become `min-w-9`, i would need to do min-w-[2.25rem] (9 / 4).
Is there a reason why I can't go unitless like `min-w-[9]` and let tailwind calculate the `2.25rem` rem value.
Otherwise we always need to add value in the tailwind.config.js
